proxy/envoy/discovery.go

```diff
@@ -15,24 +15,55 @@
 package envoy
 
 import (
+	"encoding/json"
 	"fmt"
 	"net/http"
 	"net/http/pprof"
 	"strconv"
+	"sync"
+	"sync/atomic"
 
 	restful "github.com/emicklei/go-restful"
 	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
 
 	proxyconfig "istio.io/api/proxy/v1/config"
 	"istio.io/manager/model"
 )
 
 // DiscoveryService publishes services, clusters, and routes for all proxies
 type DiscoveryService struct {
-	services model.ServiceDiscovery
-	config   *model.IstioRegistry
-	mesh     *proxyconfig.ProxyMeshConfig
-	server   *http.Server
+	services   model.ServiceDiscovery
+	controller model.Controller
+	config     *model.IstioRegistry
+	mesh       *proxyconfig.ProxyMeshConfig
+	server     *http.Server
+
+	disableCache bool
+	cacheMu      sync.RWMutex
+	cache        map[string]*discoveryCacheEntry
+
+	// TODO Profile and optimize cache eviction policy to avoid
+	// flushing the entire cache when any route, service, or endpoint
+	// changes. An explicit cache expiration policy should be
+	// considered with this change to avoid memory exhaustion as the
+	// entire cache will no longer be periodically flushed and stale
+	// entries can linger in the cache indefinitely.
+}
```

Comment:
I would add a TODO to shard the cache such that we have two caches, one for RDS/CDS (based on route rules) and another for (CDS?)SDS based on kube registry.